### ISSUE TITLE
forwarding states not restored when removing masquerade

### ISSUE THREAD

**Issue by m5aquib** [firewalld/firewalld#1122]
### What happened

firewalld does not restore network forwarding states when removing masquerade for a zone (CASE I below) and enables forwarding when removing masquerade for a zone even if it was deliberately disabled before that through sysctl (CASE II below).

### How to reproduce it (as minimally and precisely as possible)

# General configuration and initial state:

$ rpm -q firewalld
firewalld-1.3.2-1.1.noarch

$ ip -o link show | sed -e 's/<.*//g'
1: lo:
2: enp0s5

$ grep "net." /etc/sysctl.conf
net.ipv4.ip_forward = 0
net.ipv4.conf.enp0s5.forwarding = 0
net.ipv4.tcp_syncookies = 1
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.all.disable_ipv6 = 1

$ find /proc/sys/net/ipv4/conf/ -type f -name forwarding -print -exec cat {} \;
/proc/sys/net/ipv4/conf/all/forwarding
0
/proc/sys/net/ipv4/conf/default/forwarding
0
/proc/sys/net/ipv4/conf/enp0s5/forwarding
0
/proc/sys/net/ipv4/conf/lo/forwarding
0

$ firewall-cmd --list-all
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: enp0s5
  sources:
  services: dhcpv6-client ssh
  ports:
  protocols:
  forward: yes
  masquerade: no
  forward-ports:
  source-ports:
  icmp-blocks:
  rich rules:

# Steps to Reproduce:

CASE I: Add masquerade for one firewalld zone, then remove it

$ firewall-cmd --add-masquerade
success

$ firewall-cmd --list-all
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: enp0s5
  sources:
  services: dhcpv6-client ssh
  ports:
  protocols:
  forward: yes
  masquerade: yes
  forward-ports:
  source-ports:
  icmp-blocks:
  rich rules:

$ find /proc/sys/net/ipv4/conf/ -type f -name forwarding -print -exec cat {} \;
/proc/sys/net/ipv4/conf/all/forwarding
1
/proc/sys/net/ipv4/conf/default/forwarding
1
/proc/sys/net/ipv4/conf/enp0s5/forwarding
1
/proc/sys/net/ipv4/conf/lo/forwarding
1

$ firewall-cmd --remove-masquerade
success

$ firewall-cmd --list-all | grep masquerade
  masquerade: no

$ find /proc/sys/net/ipv4/conf/ -type f -name forwarding -print -exec cat {} \;
/proc/sys/net/ipv4/conf/all/forwarding
1
/proc/sys/net/ipv4/conf/default/forwarding
1
/proc/sys/net/ipv4/conf/enp0s5/forwarding
1
/proc/sys/net/ipv4/conf/lo/forwarding
1

EXPECTED: All "0"
ACTUAL: All "1"

=======================================

CASE II:

Create a state in which:
- masquerade: yes
- forwarding forcefully disabled (thorugh sysctl)

(steps continue directly from above)

$ firewall-cmd --add-masquerade
success

$ firewall-cmd --list-all | grep masquerade
  masquerade: yes

$ find /proc/sys/net/ipv4/conf/ -type f -name forwarding -print -exec cat {} \;
/proc/sys/net/ipv4/conf/all/forwarding
1
/proc/sys/net/ipv4/conf/default/forwarding
1
/proc/sys/net/ipv4/conf/enp0s5/forwarding
1
/proc/sys/net/ipv4/conf/lo/forwarding
1

$ sysctl -p | grep net
net.ipv4.ip_forward = 0
net.ipv4.conf.enp0s5.forwarding = 0
net.ipv4.tcp_syncookies = 1
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.all.disable_ipv6 = 1

$ find /proc/sys/net/ipv4/conf/ -type f -name forwarding -print -exec cat {} \;
/proc/sys/net/ipv4/conf/all/forwarding
0
/proc/sys/net/ipv4/conf/default/forwarding
0
/proc/sys/net/ipv4/conf/enp0s5/forwarding
0
/proc/sys/net/ipv4/conf/lo/forwarding
0

$ firewall-cmd --remove-masquerade
success

$ firewall-cmd --list-all | grep masquerade
  masquerade: no

$ find /proc/sys/net/ipv4/conf/ -type f -name forwarding -print -exec cat {} \;
/proc/sys/net/ipv4/conf/all/forwarding
1
/proc/sys/net/ipv4/conf/default/forwarding
1
/proc/sys/net/ipv4/conf/enp0s5/forwarding
1
/proc/sys/net/ipv4/conf/lo/forwarding
1

EXPECTED: All "0"
ACTUAL: All "1"

### Firewalld Version

1.3.2

### Firewalld Backend

nftables

### Linux distribution

openSUSE Tumbleweed

### Linux kernel version

6.3.1-1-default

**Comment by arvidjaar** [comment]
I am not convinced it is a bug. When doing masquerading firewalld must enable forwarding, otherwise the configuration will not work. But when disabling forwarding, firewalld has no information who enabled forwarding. I suppose one may store the current value when starting firewalld, but there is still corner case wen user decided to do `echo 1 > /proc/sys/net/ipv4/conf/all/forwarding` **after** firewalld has been started.

**Comment by erig0** [comment]
Agree. It's really difficult to know who changed systctls and when. I'm not sure this is worth fixing, because as @arvidjaar points out there are still corner cases.

**Comment by m5aquib** [comment]
Okay but what about the second case where removing masquerading through firewalld causes forwarding to enable when forwarding was explicitly disabled by the user?
